### PR TITLE
[cryptography/bls12381] Remove Bisection Recursion

### PR DESCRIPTION
**Pull request by patrick-ogrady**
Resolves: #977

## Summary
- replace recursive bisection in partial verification with iterative loops
- expand test coverage for single signature and boundary cases

### CODE REVIEW

File: cryptography/src/bls12381/primitives/ops.rs [hunk 1]
```diff
@@ -250,99 +250,64 @@ where
...
-        namespace,
-        message,
-        &aggregate_signature,
-    )
-    .is_ok()
-    {
-        if invalid.is_empty() {
-            return Ok(());
+        // Create aggregate public key and signature
+        let mut agg_pk = V::Public::zero();
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply cryptography/src/bls12381/primitives/ops.rs:277]: Clever, but complex suggestion. We can try this later if aggregation is ever a bottleneck here.

File: cryptography/src/bls12381/primitives/ops.rs [hunk 2]
```diff
@@ -250,99 +250,64 @@ where
...
+/// Verify a list of [PartialSignature]s by performing aggregate verification,
+/// performing repeated bisection to find invalid signatures (if any exist).
 ///
-/// Unlike `partial_verify_multiple_public_keys`, this function requires the public keys
-/// of all partial signatures to be precomputed (avoids a significant amount of compute
-/// evaluating each signer on the public polynomial).
-pub fn partial_verify_multiple_public_keys_precomputed<'a, V, I>(
-    polynomial: &[V::Public],
+/// TODO (#903): parallelize this
+fn partial_verify_multiple_public_keys_bisect<'a, V, VP>(
```
Comment (patrick-ogrady) [inline cryptography/src/bls12381/primitives/ops.rs:257]: Was able to unify the bisection logic (while removing recursion) ❤️
